### ISSUE TITLE
Dynamic row height not working in Grid

### ISSUE THREAD

**Issue by jamesonhill** [TanStack/virtual#428]
### Describe the bug

When cells within each row have varying size content, I would expect the height of the row (and in turn each cell) to be the height of the largest cell in the row. However, it seems that this is not working, unless I'm doing something incorrectly. I've read through several issues related to dynamically sizing rows (https://github.com/TanStack/virtual/issues/376, https://github.com/TanStack/virtual/issues/235, https://github.com/TanStack/virtual/issues/28) but they are all either out-dated, closed without resolution, or not quite the same use case. In the codesandbox example, i'm virtualizing the columns but forcing them all to render in order to illustrate this issue. You can see that each cell within the row is measured correctly according to the content, but I think the height of each cell should be adjusted to account for the maximum cell height, right?

### Your minimal, reproducible example

https://codesandbox.io/s/window-scroll-virtualized-ltfdib?file=/src/main.tsx

### Steps to reproduce

1. See the codesandbox repro

### Expected behavior

Height of each cell should be `Math.max(...cellHeights)`

### How often does this bug happen?

Every time

### Platform

- OS: macOs
- Browser: Chrome 107

### tanstack-virtual version

3.0.0-beta.23

### Terms & Code of Conduct

- [X] I agree to follow this project's Code of Conduct
- [X] I understand that if my bug cannot be reliable reproduced in a debuggable environment, it will probably not be fixed and this issue may even be closed.

**Comment by jamesonhill** [comment]
It looks like `virtualized.measurementsCache` does have the correct row height, but  if you apply `style={{ height: row.size }}` to the row DOM node it will break and always be the return of `virtualizer.estimateSize`.

**Comment by piecyk** [comment]
Yes, it will not take the max size but the latest measured, but this we can handle with custom `measureElement` 

```tsx
const measureElementMaxSize = <T extends unknown>(
  element: HTMLElement,
  instance: Virtualizer<T, HTMLElement>
) => {
  const index = instance.indexFromElement(element);

  const prevSize = instance.measurementsCache[index].size;
  const size = measureElement(element, instance);

  return Math.max(prevSize, size);
};
```

something like that, https://codesandbox.io/s/window-scroll-virtualized-forked-043k3x

**Comment by jamesonhill** [comment]
> Yes, it will not take the max size, only the latest measured, we can handle this case with custom `measureElement`
> 
> ```tsx
> const measureElementMaxSize = <T extends unknown>(
>   element: HTMLElement,
>   instance: Virtualizer<T, HTMLElement>
> ) => {
>   const index = instance.indexFromElement(element);
> 
>   const prevSize = instance.measurementsCache[index].size;
>   const size = measureElement(element, instance);
> 
>   return Math.max(prevSize, size);
> };
> ```
> 
> something like https://codesandbox.io/s/window-scroll-virtualized-forked-043k3x

Thank you for the suggestion. Should the codesandbox you linked work or is it just a starting point? It appears to have the same behavior. 

![Screen Shot 2022-11-18 at 1 14 36 PM](https://user-images.githubusercontent.com/25965865/202774623-fe133853-d17d-43f8-8502-602b0a1191fb.png)

**Comment by jamesonhill** [comment]
My hope was to achieve something like so:
```
// given these cell measurements per row
[[20, 50, 100], [30,20, 50], [120, 150, 100]]

// desired output to virtualizer.getVirtualItems()
[{ key: 0, index: 0, size: 100 }, { key: 1, index: 1, size: 50 }, { key: 2, index: 2, size: 150}]
```

**Comment by piecyk** [comment]
Don't know what are you trying to do but would keep one grid implementation something like https://codesandbox.io/s/window-scroll-virtualized-forked-043k3x?file=/src/main.tsx, and do the condition on cell level if you want to have special style for row 0

**Comment by jamesonhill** [comment]
![Screen Shot 2022-11-18 at 1 50 11 PM](https://user-images.githubusercontent.com/25965865/202780444-ff0d7f6d-e078-4f01-8a49-b3016e3e30c0.png)
This screenshot is an example of the layout i'm trying to achieve. Every cell per row has the same height and it uses the height of the tallest cell in the row.

**Comment by piecyk** [comment]
Ok, i think the latest example should help 👍

**Comment by jamesonhill** [comment]
Is there a better way than what I'm doing below? It seems like it _mostly_ works, but I worry it could have scaling issues. Would be nice if it could be done internally by the library.
```
const measurementsCache = useRef<Record<number, number>>({});

  const measureElementMaxSize = useCallback(
    (element: HTMLElement, instance: Virtualizer<Window, HTMLElement>) => {
      const index = instance.indexFromElement(element);

      const prevSize = instance.measurementsCache[index].size;
      const size = measureElement(element, instance);
      const newSize = Math.max(prevSize, size);
      if (
        !measurementsCache.current[index] ||
        newSize > measurementsCache.current[index]
      ) {
        measurementsCache.current[index] = newSize;
      }

      return newSize;
    },
    []
  );

// on cell renderer node
style={{ height: measurementsCache.current[index] }}
```

Here's a demo https://codesandbox.io/s/window-scroll-virtualized-ltfdib?file=/src/main.tsx

**Comment by jamesonhill** [comment]
Hmm, actually after further testing my path above may not quite work either.

**Comment by piecyk** [comment]
> Is there a better way than what I'm doing below? 

I think this is fine, you can always extract it to custom hook.

> Hmm, actually after further testing my path above may not quite work either.

Internally we have a optimization to recalculate position from index where we had new size. In case of max size we need to start from last index where the last size was changed, for simplification i just start from 0.
 
```
// recalculate position from start
(instance as any).pendingMeasuredCacheIndexes.push(0);
```

Also in case where we have text in cell, the width need to be fixed, commented out ref for columns as the width is known.

https://codesandbox.io/s/window-scroll-virtualized-forked-mxv603?file=/src/main.tsx

I hope this helps 👍

**Comment by piecyk** [comment]
@jamesonhill maybe go with a bit different approach, first position rows, then render columns as one block, thanks to     minHeight: row.size + ResizeObserver there is no need for custom measureElementMaxSize to store max size.

Checkout this https://codesandbox.io/s/window-scroll-virtualized-forked-pzebvz?file=/src/main.tsx

**Comment by jamesonhill** [comment]
@piecyk Brilliant. Thanks a ton for your help.  This could be a good example to add to the documentation. Maybe i'll add a PR for it, but will close this issue.

**Comment by kavyantic** [comment]
I read the whole thread patiently, including the existing threads related to this issue. I am facing a similar problem in `useVirtualizer ` hook where I do not know why measureElement is never triggered.
```
const rowVirtualizer = useVirtualizer({
    count: rows.length,
    estimateSize: () => 60, //estimate row height for accurate scrollbar dragging
    getScrollElement: () => tableContainerRef.current,
    measureElement:
      (element) => {
        console.log("measuring element");
        return element?.getBoundingClientRect().height;
      },
    overscan: 5,
  });
```
```
<tr ref={(node) => rowVirtualizer.measureElement(node)} > {content} </tr>
```

I tested it many times but measureElement function is not called even once and virtualizer uses estimated size for all rows.

**Comment by jamesonhill** [comment]
@kavyantic  I'm a few versions behind, but I don't use `measureElement` on `useVirtualizer`. I do `<div ref={rowVirtualizer.measureElement} ...otherStuff` and it works great.

**Comment by kavyantic** [comment]
After thoroughly looking at the example provided  [here](https://tanstack.com/virtual/latest/docs/framework/react/examples/dynamic) I realized that I was missing a  `data-index={virtualRow.index}` attribute on `<tr/>` element which is required for virtualizer to provide accurate translateY pixels to the row element. Now it is dynamically adjusting row translate respective it's height as expected. Thanks for acknowledgement and Big thanks! to the TanStack team for such a powerfull package. I used to use `@tanstack/react-query` for qureying backend and thought I should try more packages from Tanstack after the new documentation release and now I can't avoid using:-

- @tanstack/react-table
- @tanstack/react-router
- @tanstack/react-virtual
- @tanstack/react-query
- @tanstack/react-form

All of these packages are great.
Specially the `@tanstack/react-table` in combination with `@tanstack/react-virtual` is a game changer.
